Makefile.include

```diff
@@ -69,32 +69,37 @@ MAKEOVERRIDES += $(foreach v,$(__OVERRIDE_DIRECTORY_VARIABLES),$(v)=$($(v)))
 
 # Deprecation of configuring 'RIOTBOARD'
 ifneq ($(abspath $(RIOTBASE)/boards),$(abspath $(RIOTBOARD)))
-  $(warning overriding RIOTBOARD for external boards is deprecated, please use an absolute BOARDSDIR)
+  $(warning overriding RIOTBOARD for external boards is deprecated, use EXTERNAL_BOARD_DIRS instead)
   override RIOTBOARD      := $(abspath $(RIOTBOARD))
   __DIRECTORY_VARIABLES += RIOTBOARD
 endif
 
-# Make all paths absolute.
-override RIOTBASE       := $(abspath $(RIOTBASE))
-override RIOTCPU        := $(abspath $(RIOTCPU))
-override RIOTMAKE       := $(abspath $(RIOTMAKE))
-override RIOTPKG        := $(abspath $(RIOTPKG))
-override RIOTTOOLS      := $(abspath $(RIOTTOOLS))
-override RIOTPROJECT    := $(abspath $(RIOTPROJECT))
-override GITCACHE       := $(abspath $(GITCACHE))
-override APPDIR         := $(abspath $(APPDIR))
-override BUILD_DIR      := $(abspath $(BUILD_DIR))
-override BINDIRBASE     := $(abspath $(BINDIRBASE))
-override BINDIR         := $(abspath $(BINDIR))
-override PKGDIRBASE     := $(abspath $(PKGDIRBASE))
-override DLCACHE_DIR    := $(abspath $(DLCACHE_DIR))
-
-# Keep standard make behavior for new variables
-__DIRECTORY_VARIABLES += \
-  BOARDSDIR \
-  #
+ifneq (,$(BOARDSDIR))
+  # Only warn users, not the CI.
+  ifneq ($(RIOT_CI_BUILD),1)
+    $(warning Using BOARDSDIR is deprecated use EXTERNAL_BOARD_DIRS instead)
+    $(info EXTERNAL_BOARD_DIRS can contain multiple folders separated by space)
+  endif
+  EXTERNAL_BOARD_DIRS += $(BOARDSDIR)
+endif
```

Comment:
On my final check I just noticed this. I think its better to avoid `override`, and have it be a normal variable. But I won't request this change, address if you want.

Reply:
Didn't select the right lines, its more about L100-102

Reply:
I think this is needed. E.g. if I would run

```
EXTERNAL_BOARD_DIRS=../extra_boards make BOARD=foo -C examples/default
```

the relative path in `EXTERNAL_BOARD_DIRS` would be replaced by the absolute path. But when running

```
make BOARD=foo EXTERNAL_BOARD_DIRS=../extra_boards -C examples/default
```

without the `override`, `EXTERNAL_BOARD_DIRS` will continue to contain a relative path. This is bound to cause issues during compilation.

Reply:
> without the `override`, `EXTERNAL_BOARD_DIRS` will continue to contain a relative path. This is bound to cause issues during compilation.

Yes that is correct, it is necessary to keep that usage, but you are making variables passed after/before `make` to always behave as environment variables and not as make arguments, which is forcing make to behave in a non standard way.

Also since noticing this I looked more into the code (and make doc) and `override` variables are not passed to `sub-makes`, even if exported so this can all lead to weird to debug situation. For the other directory variables we add them to `MAKEOVERRIDES` to force this. I'm not sure if this could be an issue, but I think for consistency there are two choices:

- add `EXTERNAL_BOARD_DIRS` to `MAKEOVERRIDES`
- have `EXTERNAL_BOARD_DIRS` as a normal make variable (don't `override`)

I personally prefer option 2, because all this `override` and `MAKEOVERRIDES` feels a a bit hacky to me, and from what I have read adding to `MAKEOVERRIDES` is not recommended, I would rather require proper make usage. But since the other variables use this I think its your choice to make as the contributor.

The following thread explains it well IMO https://stackoverflow.com/questions/326\37204/why-doesnt-gnu-makes-override-pass-through-to-sub-makes.

Reply:
> but I think for consistency there are two choices:

Maybe leaving it as is could be a good choice as well, but since its exported it means there is an expectation of it being used in a sub-make, and if there is such expectation than the `override` issue I mentioned could come to happen.

Reply:
All the uses of `override` + `export`  are starting to look bad to me right now.

Reply:
I don't think this is an issue here, as `Makefile.include` is always included. So even though the sub-makes initially will have the non-absolute path again, they will convert it to an absolute path again.

Quite nasty and hacky, but it should work reliable.

Reply:
> I don't think this is an issue here, as Makefile.include is always included. So even though the sub-makes initially will have the non-absolute path again, they will convert it to an absolute path again.
> 
> Quite nasty and hacky, but it should work reliable.

OK I think you are right, it shouldn't lead to an issue here (I tried to get one playing around with normal usage of `EXTERNAL_BOARD_DIRS`). An the case you referenced https://github.com/RIOT-OS/RIOT/pull/13862#discussion_r414552957 does lead to an error on `tests/riotboot` if there is no override.

Style wise I still prefer not overriding make behaviour, but its already what we are doing for other variables and it is needed If we want to have `EXTERNAL_BOARD_DIRS` passed correctly as a make command.

Reply:
I just remembered why I didn't add `EXTERNAL_BOARD_DIRS` to `__DIRECTORY_VARIABLES`: There is a check that enforces them to be neither empty nor contain spaces. Both can happen to `EXTERNAL_BOARD_DIRS`. Having it added to `MAKEOVERRIDES` would be consistent, but doing so without using the `__DIRECTORY_VARIABLES` approach seems to be equally nasty than just relying on `Makefile.include` turning path in `EXTERNAL_BOARD_DIRS` again into an absolute path.

Maybe this would be the least nasty approach?

```patch
diff --git a/Makefile.include b/Makefile.include
index 1f06f18403..c8eb553825 100644
--- a/Makefile.include
+++ b/Makefile.include
@@ -63,9 +63,13 @@ __DIRECTORY_VARIABLES := \
 
 # In GNU make command line is supposed to override the value set in makefile
 # These variables can be overridden even when set from command line
-__OVERRIDE_DIRECTORY_VARIABLES := $(__DIRECTORY_VARIABLES)
+__OVERRIDE_VARIABLES := $(__DIRECTORY_VARIABLES)
+# Also override EXTERNAL_BOARD_DIRS, but do not perform the sanity checks on
+# it (opposed to entries in __DIRECTORY_VARIABLES), as it can be empty and
+# spaces are used as separator between directories
+__OVERRIDE_VARIABLES += EXTERNAL_BOARD_DIRS
 # Use absolute paths in recursive "make" even if overridden on command line.
-MAKEOVERRIDES += $(foreach v,$(__OVERRIDE_DIRECTORY_VARIABLES),$(v)=$($(v)))
+MAKEOVERRIDES += $(foreach v,$(__OVERRIDE_VARIABLES),$(v)=$($(v)))
 
 # Deprecation of configuring 'RIOTBOARD'
 ifneq ($(abspath $(RIOTBASE)/boards),$(abspath $(RIOTBOARD)))
```

Reply:
I think I would rather not mess with `MAKEOVERRIDES `,  and https://stackoverflow.com/questions/32637204/why-doesnt-gnu-makes-override-pass-through-to-sub-makes they mentioned issues with whitespaces.

Reply:
So I though about this a little more over the weekend.

I still prefer not supporting passing relative paths as command line arguments, so not to use override or `MAKEOVERRIDES`, and just document that it only works this way (force to properly use make instead of hacking over it). But, again this is not the current behavior most RIOT user expect, so that had me re-thinking your statement:

> be equally nasty than just relying on Makefile.include turning path i EXTERNAL_BOARD_DIRS again into an absolute path.

I think you are right here, trusting something that is not explicit is IMO worst than trusting an explicit HACK. So lets consider the explicit hack. Regarding your suggested patch, we could add it on its own line with the comments you included:

```mk
# In GNU make command line is supposed to override the value set in makefile
# These variables can be overridden even when set from command line
__OVERRIDE_DIRECTORY_VARIABLES := $(__DIRECTORY_VARIABLES)
# Use absolute paths in recursive "make" even if overridden on command line.
MAKEOVERRIDES += $(foreach v,$(__OVERRIDE_DIRECTORY_VARIABLES),$(v)=$($(v)))
# Also override EXTERNAL_BOARD_DIRS, but do not perform the sanity checks on
# it (opposed to entries in __DIRECTORY_VARIABLES), as it can be empty and
# spaces are used as separator between directories
MAKEOVERRIDES += EXTERNAL_BOARD_DIRS=$(EXTERNAL_BOARD_DIRS)
```

Ok, but will white spaces be an issue? Seem like they would need to be escaped, so something like the following:

```
MAKEOVERRIDES += EXTERNAL_BOARD_DIRS=$(subst $(space),\ ,$(EXTERNAL_BOARD_DIRS))
```

Maybe also adding a `$(strip`

`MAKEOVERRIDES` from everything I have read seems really tricky, since it actually only takes effect if a variable has been set from the command line, otherwise you would also need an export statement. In this case we always export these directory variables, so it seems everything is in place for it to work.

So clearly a lot of conditions need to happen for this `MAKEOVERRIDE` and `override` to work well. A lot of hacks/workarounds are needed, but at least explicit and we are not relying on `Makefile.include` converting the path.

So lets list our options (I'll list them in my order of preference):

- Don't use `override` or `MAKEOVERRIDE`, document that `EXTERNAL_BOARD_DIRS` passed through the command line should have absolute paths.
- User `override` but document why we don't add it to `MAKEOVERRIDE` and why we trust it to work
- Document and put in place the list of workarounds mentioned above.

@maribu I would be OK with any of the above 3, so again I leave it up to you, but whatever we do, lets just add a big comment explaining it. I don't expect you to document the `MAKEOVERRIDE` behaviour, its not related to this PR.

Reply:
Maybe we could also just check if `EXTERNAL_BOARD_DIRS` is given as command line argument and abort if so?

Reply:
Something like
``` Makefile
ifeq ($(origin EXTERNAL_BOARD_DIRS),command line)
  $(error EXTERNAL_BOARD_DIRS must be passed as environment variable, and not as command line argument)
endif
```

Reply:
Sound good to me!